datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -275,6 +275,93 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
     })
 }
 
+fn date_bin_single(stride: i64, source: i64, origin: i64) -> i64 {
+    let time_diff = source - origin;
+    // distance to bin
+    let time_delta = time_diff - (time_diff % stride);
+
+    let time_delta = if time_diff < 0 && stride > 1 {
+        // The origin is later than the source timestamp, round down to the previous bin
+        time_delta - stride
+    } else {
+        time_delta
+    };
+
+    origin + time_delta
+}
+
+/// DATE_BIN sql function
+pub fn date_bin(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    if args.len() != 3 {
+        return Err(DataFusionError::Execution(
+            "Expected three arguments for DATE_BIN".to_string(),
+        ));
+    }
+
+    let (stride, array, origin) = (&args[0], &args[1], &args[2]);
+
+    let stride = match stride {
+        ColumnarValue::Scalar(ScalarValue::IntervalDayTime(Some(v))) => {
+            let (days, ms) = IntervalDayTimeType::to_parts(*v);
+            let nanos = (Duration::days(days as i64) + Duration::milliseconds(ms as i64))
+                .num_nanoseconds();
+            match nanos {
+                Some(v) => v,
+                _ => {
+                    return Err(DataFusionError::Execution(
+                        "stride of DATE_BIN is too large".to_string(),
+                    ))
+                }
+            }
+        }
+        _ => {
+            return Err(DataFusionError::Execution(
+                "stride of DATE_BIN is an invalid type".to_string(),
+            ))
+        }
+    };
+
+    let origin = match origin {
```

Comment:
Same comment applies here as about stride above -- I recommend making the error message more explicit as there is a difference between the type of constant (which the user can possibly do something about) and trying to use an array (which the user can not do anything about)